File: 5kyu/Rot13/javascript/main.js
function rot13(message) {
    return message.split('')
        .map(word => {
            if ((/[a-m]/gi).test(word)) {
                // is A~M or a~m
                return String.fromCharCode(word.charCodeAt(0) + 13);
            } else if ((/[n-z]/gi).test(word)) {
                // is N~Z or n~z
                return String.fromCharCode(word.charCodeAt(0) - 13);
            } else {
                // is not latin alphabet
                return word
            }
        }).join('');
}

File: 5kyu/Rot13/rust/main.rs
fn rot13(message: &str) -> String {
    let rot_bytes: Vec<u8> = message
        .as_bytes()
        .iter()
        .map(|x| match x {
            b'a'..=b'm' => x + 13,
            b'A'..=b'M' => x + 13,
            b'n'..=b'z' => x - 13,
            b'N'..=b'Z' => x - 13,
            _ => *x,
        })
        .collect::<Vec<u8>>();

    std::str::from_utf8(&rot_bytes).unwrap().to_string()
}

#[cfg(test)]
mod tests {
    use super::rot13;

    const ERR_MSG: &str = "\nYour result (left) did not match the expected output (right)";

    fn dotest(s: &str, expected: &str) {
        assert_eq!(rot13(s), expected, "{ERR_MSG} with message = \"{s}\"")
    }

    #[test]
    fn sample_tests() {
        dotest("test", "grfg");
        dotest("Test", "Grfg");
    }
}
